plutus-errors/src/TH/Codes.hs

```diff
@@ -0,0 +1,20 @@
+{-# LANGUAGE TemplateHaskell #-}
+module TH.Codes (genCodes) where
+
+import Language.Haskell.TH
+import Language.Plutus.Common
+
+genCodes :: [Name] -> Q Exp
+genCodes ns = listE $ fmap (\ n -> do
+   DataConI n' t _  <- reify n
+   fun <- [|errorCode|]
+   pure $ AppE fun $ genSaturatedDataCon n' t
+   ) ns
+
+genSaturatedDataCon :: Name -> Type -> Exp
+genSaturatedDataCon = genSaturatedDataCon' . ConE
+    where
+      genSaturatedDataCon' acc (ForallT _ _ t) = genSaturatedDataCon' acc t
+      genSaturatedDataCon' acc (AppT _ (VarT _)) = acc
+      genSaturatedDataCon' acc (AppT _ t') = genSaturatedDataCon' (acc `AppE` VarE (mkName "Prelude.undefined")) t'
```

Comment:
This looks wrong. `AppT` is type applications, like `T a`, which don't correspond to value-level arguments. *Functions* correspond to value-level arguments, i.e. `ArrowT`.

Reply:
I suggest using `th-abstraction`, which deals with the details of getting constructor arguments etc. for you.